filmicrgb: introduce gamut mapping at output

flannelhead opened the pull request:
Draft, OpenCL not done yet. Testing and comments are welcome. Pinging @aurelienpierre, you might want to have a look.

Introduce a new color science version to Filmic RGB that compresses the output to the target gamut. This replaces the desaturation curve that is used in the current version.

This brings a couple of advantages. In the examples, left is current darktable output and right is with the proposed changes. Only difference is the selected color science version.

1. If the target color space gamut is smaller than the working space gamut, colors are gracefully compressed to the target gamut, preserving hue and details in highly saturated regions.
![gamut_mapping](https://user-images.githubusercontent.com/5001906/150700543-7b15392b-e2b9-43d0-bea7-aee7608b7003.png)
2. Preservation of overall colorfulness, avoiding washed out blue skies and allowing to reach e.g. bright yellows easier.
![gamut_mapping_2](https://user-images.githubusercontent.com/5001906/150700621-3c3bbd57-c7d1-4fa5-8d44-b787be07791d.png)

In these examples, target color space is sRGB. This is currently hardcoded in this draft PR. I also think this should be the default choice because filmic is usually directly followed by conversion to export profile which is usually sRGB. It would be also good if there was a choice "export profile" which would use any export profile that is currently set in the colorout module.

In a nutshell:

* gamut compression takes place in [Oklab perceptual color space](https://bottosson.github.io/posts/oklab/) (LCh cylindrical representation)
* gamut boundary is approximated by polynomials fitted to the upper and lower boundaries at each hue slice
* pixels are compressed towards the neutral axis at constant lightness
* compression uses a soft knee that is configurable at the `look` tab in filmic when you select the v6 color science. 1 is hard clipping, 0 rolls all the pixels through the soft knee, resulting in a gentle look

This is working quite well for me so far.

Please note that this PR contains three commits that are meant to be merged in another PR, https://github.com/darktable-org/darktable/pull/10963. I just wanted to base my work here on those changes.

Open items / TODO:

* the target color space should be selectable in filmic GUI
* experimenting with Gaussian blurring of the gamut boundary instead of the current box filter
* experimenting with other perceptual color spaces / CAMs such as CAM18sl
* OpenCL version

aurelienpierre commented:
Right, I have spent the past 2 days on filmic v6 with gamut mapping too, but in Ych 🤣

aurelienpierre commented on src/common/gamut_mapping.c:
```diff
@@ -0,0 +1,333 @@
+
+static float find_max_C(const float a, const float b, const float L, const float min_C, const float black_luminance, const dt_colormatrix_t output_matrix)
+{
+  // Find the lower boundary maximum chroma at given lightness by bisection
```
Given that the gamut is a double cone, its boundary can be modelled by saturation = chroma / lightness for the bottom part. 
Then, for the upper part, you find the line of (chroma / (white - lightness)).

flannelhead replied:
In general the gamut is not necessarily a simple double cone constrained by straight lines at each slice (for e.g. JzAzBz, both upper and lower boundary are curved). Therefore I intentionally made the gamut boundary handling a little more generic than it needs to be for Oklab, to allow for experimentation with other color spaces.

For Oklab the lower boundary seems to be a straight line but the upper one is slightly curved, therefore the 5th degree polynomial.
![gamut_slice](https://user-images.githubusercontent.com/5001906/150766635-b64d4936-04d2-4be4-b886-86d5c7845a11.png)

See also @bottosson's excellent [post](https://bottosson.github.io/posts/gamutclipping/#gamut-clipping) where there are more images of hue slices in Oklab.

aurelienpierre replied:
Ok thanks

aurelienpierre commented:
So, I understand you went to the perceptual colorimetry, while I went down the relative colorimetric intent. I think we can merge both changes since the difference between our codes is mostly the last gamut mapping function.

aurelienpierre commented on src/iop/channelmixerrgb.c:
```diff
@@ -687,7 +687,7 @@ static inline void loop_switch(const float *const restrict in, float *const rest
       {
         // Convert from RGB to XYZ
-        dot_product(temp_two, RGB_to_XYZ, temp_one);
+        dt_apply_transposed_color_matrix(temp_two, RGB_to_XYZ, temp_one);
```
why ?

johnny-bit replied:
I think the dot_product -> dt_apply_transposed_color_matrix stem from https://github.com/darktable-org/darktable/pull/10963 and that pr has all the explaination for it.

flannelhead replied:
Yep, as I had to deal with color matrices here and already had the optimizations of https://github.com/darktable-org/darktable/pull/10963 done but not merged yet, I included the commits here as well. These changes can be safely ignored.

flannelhead commented:
> Right, I have spent the past 2 days on filmic v6 with gamut mapping too, but in Ych

I also started mine just last week :rofl: Well, isn't it nice to have two independent implementations to compare. Possibly they have different strengths. :)

> So, I understand you went to the perceptual colorimetry, while I went down the relative colorimetric intent. I think we can merge both changes since the difference between our codes is mostly the last gamut mapping function.

By perceptual do you mean that I used the soft knee and you seem to hard-clip to the gamut boundary (or actually iterate toward that)?

Indeed that's a difference, and the color spaces are different too - in your method there's luminance Y and the Kirk chroma, while I used Oklab's lightness L and chroma. It will be interesting to see how these compare in practice.

Also note that I map currently directly to sRGB gamut, but I think the output color profile would be a sensible choice there. Your implementation currently clips to pipe RGB but I see you are planning to clip to output profile.

If these both are going to be merged, I would be happy to rebase my work on yours. Finishing this one might take me a while so yours is more likely to be merged first.

aurelienpierre commented:
> By perceptual do you mean that I used the soft knee and you seem to hard-clip to the gamut boundary (or actually iterate toward that)?

No, I thought that you contracted the whole space to fit within the gamut. But that would require a separate loop to get the max chroma = f(hue), then only a mapping loop.

flannelhead commented:
> No, I thought that you contracted the whole space to fit within the gamut. But that would require a separate loop to get the max chroma = f(hue), then only a mapping loop.

I think my code effectively does this - the gamut boundary of target color space is modeled, and source pixel chroma is reduced at constant hue and lightness to fit inside the target gamut. Currently a soft clipping exponential function is used so in theory it supports an infinitely wide source gamut, and the soft knee position is adjustable to spread the out-of-target gamut colors to a wider or narrower range. But it could be also possible to define a more accurate mapping if the source gamut boundary would be also precalculated.

aurelienpierre commented:
> Currently a soft clipping exponential function is used so in theory it supports an infinitely wide source gamut, and the soft knee position is adjustable to spread the out-of-target gamut colors to a wider or narrower range. But it could be also possible to define a more accurate mapping if the source gamut boundary would be also precalculated.

Yes, I think the knee parameter shouldn't be exposed to the user but computed from something that scales with the gamut, because otherwise it's impossible to predict the appearance of the gamut mapping in other spaces. Also, mapping flatly like that has the drawback of remapping even when there is no need to.

flannelhead commented:
There's now remapping from the source gamut to target gamut. There's a scaling: pixels with chroma below knee (90 % of the maximum available chroma at given hue and lightness) are not touched; then the range from the knee to source gamut boundary is remapped to the remaining available 10 %. The knee is now hardcoded and not exposed to the user. The filmic saturation control is not yet enabled, but it should be handled when this is rebased on Aurélien's work.

flannelhead commented:
This PR has been given a lot of polish.

* Removed the dependency on the other PR (related to `dot_product` and `dt_apply_transposed_color_matrix()`).
* Added caching of the gamut boundary data. The Filmic RGB instance doesn't have to care about lifecycle management at all and the data will be reusable across instances.
* Improved smoothing and accuracy of the gamut boundary.

OpenCL is missing but I will do it only if / when the CPU path is merged along with Aurelien's gamut mapping method. Some checks for validity of the output profile are required, but otherwise this one should be ready for review.

TurboGit commented:
@flannelhead : Given that Filmic V6 is now merged and that it has introduced a gamut mapping too... What is the status of this PR? TIA.

flannelhead commented:
@TurboGit : I intend to rebase this on master at some point for easier evaluation of the differences between this and the merged approach. However I think the merged approach ended in a very good shape with less code, and probably there are some things to solve (such as handling of highlights) in general with the v6 color science. I'd prefer not to have this merged for 4.0 but instead first see how the merged v6 takes off.

flannelhead commented:
Oops, didn't mean to close it... Instead I'll convert this to a draft.

RawConvert commented:
@flannelhead , I like what your method does in the two examples at the top of this issue, and sometimes I have washed-out blue skies myself.  Obviously code has moved on since then.  Is it possible to repeat the examples showing current V6 filmic as compared to your current state-of-play?

flannelhead commented:
@RawConvert, thanks for the interest. Currently I don't have sufficient time for the rebase or comparisons (focusing on actual photography for a while now ;) However I'll do it at some point. The last time I tested, I found that the two methods (filmic v6 and the one in this PR) produced 90 % similar results. The filmic v6 gamut mapping has a smaller code footprint than this one (and this one doesn't even have OpenCL yet) and it is more straightforward.

I encourage you to experiment with it as well as the different norms. You should certainly be able to get rid of pale blue skies with that one.

RawConvert commented:
Thanks, and yes, I will try it.

flannelhead commented:
I'm closing this one since there should be something better incoming quite soon.
